File: apps/wallet-desktop/desktop/src-tauri/src/services/supabase_service.rs
/// Supabase service for Desktop (Rust)
/// Mirrors iOS SupabaseService.swift functionality

use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tokio::sync::RwLock;

use crate::services::supabase_config::SupabaseConfig;

// TODO: Replace with actual Supabase Rust SDK
// For now, using HTTP client approach
use reqwest::Client;

pub struct SupabaseService {
    client: Arc<Client>,
    url: String,
    key: String,
    current_session: Arc<RwLock<Option<Session>>>,
}

impl SupabaseService {
    pub fn new() -> Self {
        Self {
            client: Arc::new(Client::new()),
            url: SupabaseConfig::supabase_url().to_string(),
            key: SupabaseConfig::supabase_key().to_string(),
            current_session: Arc::new(RwLock::new(None)),
        }
    }

    pub async fn check_session(&self) -> Result<(), Box<dyn std::error::Error>> {
        // TODO: Implement session check using Supabase Auth API
        // For now, placeholder
        Ok(())
    }

    pub async fn sign_up(
        &self,
        email: &str,
        password: &str,
    ) -> Result<Session, Box<dyn std::error::Error>> {
        let url = format!("{}/auth/v1/signup", self.url);
        
        let response = self.client
            .post(&url)
            .header("apikey", &self.key)
            .header("Content-Type", "application/json")
            .json(&serde_json::json!({
                "email": email,
                "password": password
            }))
            .send()
            .await?;
        
        if response.status().is_success() {
            let auth_response: AuthResponse = response.json().await?;
            if let Some(session) = auth_response.session {
                let session = Session {
                    user_id: session.user.id,
                    access_token: session.access_token,
                    refresh_token: session.refresh_token.unwrap_or_default(),
                };
                *self.current_session.write().await = Some(session.clone());
                return Ok(session);
            }
        }
        
        Err("Sign up failed".into())
    }

    pub async fn sign_in(
        &self,
        email: &str,
        password: &str,
    ) -> Result<Session, Box<dyn std::error::Error>> {
        let url = format!("{}/auth/v1/token?grant_type=password", self.url);
        
        let response = self.client
            .post(&url)
            .header("apikey", &self.key)
            .header("Content-Type", "application/json")
            .json(&serde_json::json!({
                "email": email,
                "password": password
            }))
            .send()
            .await?;
        
        if response.status().is_success() {
            let auth_response: AuthResponse = response.json().await?;
            if let Some(session) = auth_response.session {
                let session = Session {
                    user_id: session.user.id,
                    access_token: session.access_token,
                    refresh_token: session.refresh_token.unwrap_or_default(),
                };
                *self.current_session.write().await = Some(session.clone());
                return Ok(session);
            }
        }
        
        Err("Sign in failed".into())
    }

    pub async fn sign_out(&self) -> Result<(), Box<dyn std::error::Error>> {
        // TODO: Implement sign out
        *self.current_session.write().await = None;
        Ok(())
    }

    pub async fn register_device(
        &self,
        platform: &str,
        label: &str,
        push_token: Option<&str>,
    ) -> Result<Device, Box<dyn std::error::Error>> {
        let session = self.current_session.read().await.clone();
        let session = session.ok_or("Not authenticated")?;
        
        let url = format!("{}/rest/v1/devices", self.url);
        
        let response = self.client
            .post(&url)
            .header("apikey", &self.key)
            .header("Authorization", format!("Bearer {}", session.access_token))
            .header("Content-Type", "application/json")
            .header("Prefer", "return=representation")
            .json(&serde_json::json!({
                "user_id": session.user_id,
                "platform": platform,
                "device_label": label,
                "push_token": push_token,
                "is_active": true
            }))
            .send()
            .await?;
        
        if response.status().is_success() {
            let device: Device = response.json().await?;
            return Ok(device);
        }
        
        Err("Device registration failed".into())
    }

    pub async fn create_account(
        &self,
        chain: &str,
        address: &str,
        account_type: &str,
        is_default: bool,
    ) -> Result<SupabaseAccount, Box<dyn std::error::Error>> {
        let session = self.current_session.read().await.clone();
        let session = session.ok_or("Not authenticated")?;
        
        let url = format!("{}/rest/v1/accounts", self.url);
        
        let response = self.client
            .post(&url)
            .header("apikey", &self.key)
            .header("Authorization", format!("Bearer {}", session.access_token))
            .header("Content-Type", "application/json")
            .header("Prefer", "return=representation")
            .json(&serde_json::json!({
                "user_id": session.user_id,
                "chain": chain,
                "address": address,
                "type": account_type,
                "is_default": is_default
            }))
            .send()
            .await?;
        
        if response.status().is_success() {
            let account: SupabaseAccount = response.json().await?;
            return Ok(account);
        }
        
        Err("Account creation failed".into())
    }

    pub async fn fetch_accounts(&self) -> Result<Vec<SupabaseAccount>, Box<dyn std::error::Error>> {
        let session = self.current_session.read().await.clone();
        let session = session.ok_or("Not authenticated")?;
        
        let url = format!("{}/rest/v1/accounts?user_id=eq.{}", self.url, session.user_id);
        
        let response = self.client
            .get(&url)
            .header("apikey", &self.key)
            .header("Authorization", format!("Bearer {}", session.access_token))
            .send()
            .await?;
        
        if response.status().is_success() {
            let accounts: Vec<SupabaseAccount> = response.json().await?;
            return Ok(accounts);
        }
        
        Ok(vec![])
    }

    pub fn get_session(&self) -> Option<Session> {
        // TODO: Make async or use Arc properly
        None
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Session {
    pub user_id: String,
    pub access_token: String,
    pub refresh_token: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Device {
    pub id: String,
    pub user_id: String,
    pub platform: String,
    pub device_label: Option<String>,
    pub push_token: Option<String>,
    pub is_active: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SupabaseAccount {
    pub id: String,
    pub user_id: String,
    pub chain: String,
    pub address: String,
    #[serde(rename = "type")]
    pub account_type: String,
    pub is_default: bool,
}

#[derive(Debug, Serialize, Deserialize)]
struct AuthResponse {
    session: Option<SessionData>,
}

#[derive(Debug, Serialize, Deserialize)]
struct SessionData {
    user: UserData,
    access_token: String,
    refresh_token: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
struct UserData {
    id: String,
}

File: backup/wallets/android-wallet/app/src/main/java/com/noor/wallet/services/SupabaseService.kt
package com.nor.wallet.services

import android.util.Log
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.createSupabaseClient
import io.github.jan.supabase.gotrue.auth
import io.github.jan.supabase.gotrue.providers.builtin.Email
import io.github.jan.supabase.postgrest.from
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Supabase service for Android
 * Mirrors iOS SupabaseService.swift functionality
 */
class SupabaseService private constructor() {
    companion object {
        @Volatile
        private var INSTANCE: SupabaseService? = null

        fun getInstance(): SupabaseService {
            return INSTANCE ?: synchronized(this) {
                val instance = SupabaseService()
                INSTANCE = instance
                instance
            }
        }
    }

    private val supabase: SupabaseClient = createSupabaseClient(
        supabaseUrl = SupabaseConfig.supabaseURL,
        supabaseKey = SupabaseConfig.supabaseKey
    ) {
        install(io.github.jan.supabase.gotrue.Auth)
        install(io.github.jan.supabase.postgrest.Postgrest)
        install(io.github.jan.supabase.realtime.Realtime)
        install(io.github.jan.supabase.storage.Storage)
        install(io.github.jan.supabase.functions.Functions)
    }

    private val _currentSession = MutableStateFlow<Session?>(null)
    val currentSession: StateFlow<Session?> = _currentSession.asStateFlow()

    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()

    init {
        // Check existing session on init
        checkSession()
        observeAuthChanges()
    }

    // MARK: - Authentication

    suspend fun checkSession() {
        try {
            val session = supabase.auth.currentSessionOrNull()
            _currentSession.value = session?.let { Session(it) }
            _isAuthenticated.value = session != null
        } catch (e: Exception) {
            if (SupabaseConfig.enableDebugLogging) {
                Log.w(TAG, "Session check failed: ${e.message}")
            }
            _currentSession.value = null
            _isAuthenticated.value = false
        }
    }

    suspend fun signUp(email: String, password: String): Session {
        val response = supabase.auth.signUpWith(Email) {
            this.email = email
            this.password = password
        }
        
        val session = response.session
        if (session != null) {
            _currentSession.value = Session(session)
            _isAuthenticated.value = true
            return Session(session)
        } else {
            throw Exception("No session returned")
        }
    }

    suspend fun signIn(email: String, password: String): Session {
        val response = supabase.auth.signInWith(Email) {
            this.email = email
            this.password = password
        }
        
        val session = response.session
        if (session != null) {
            _currentSession.value = Session(session)
            _isAuthenticated.value = true
            return Session(session)
        } else {
            throw Exception("No session returned")
        }
    }

    suspend fun signOut() {
        supabase.auth.signOut()
        _currentSession.value = null
        _isAuthenticated.value = false
    }

    private fun observeAuthChanges() {
        // TODO: Implement auth state listener
        // Supabase Kotlin SDK auth state changes
    }

    // MARK: - Devices

    data class Device(
        val id: String,
        val userId: String,
        val platform: String,
        val label: String,
        val pushToken: String?
    )

    suspend fun registerDevice(
        platform: String,
        label: String,
        pushToken: String?
    ): Device {
        val userId = _currentSession.value?.userId ?: throw Exception("Not authenticated")
        
        val deviceData = mapOf(
            "user_id" to userId,
            "platform" to platform,
            "label" to label,
            "push_token" to pushToken,
            "last_seen" to System.currentTimeMillis()
        )

        val response = supabase.from("devices").insert(deviceData)
        // Parse response and return Device
        // TODO: Parse response properly
        return Device(
            id = "",
            userId = userId,
            platform = platform,
            label = label,
            pushToken = pushToken
        )
    }

    // MARK: - Accounts

    data class SupabaseAccount(
        val id: String,
        val userId: String,
        val chain: String,
        val address: String,
        val type: String,
        val isDefault: Boolean
    )

    suspend fun createAccount(
        chain: String,
        address: String,
        type: String,
        isDefault: Boolean
    ): SupabaseAccount {
        val userId = _currentSession.value?.userId ?: throw Exception("Not authenticated")
        
        val accountData = mapOf(
            "user_id" to userId,
            "chain" to chain,
            "address" to address,
            "type" to type,
            "is_default" to isDefault
        )

        val response = supabase.from("accounts").insert(accountData)
        // TODO: Parse response
        return SupabaseAccount(
            id = "",
            userId = userId,
            chain = chain,
            address = address,
            type = type,
            isDefault = isDefault
        )
    }

    suspend fun fetchAccounts(): List<SupabaseAccount> {
        val userId = _currentSession.value?.userId ?: throw Exception("Not authenticated")
        
        val response = supabase.from("accounts")
            .select {
                filter {
                    eq("user_id", userId)
                }
            }
        
        // TODO: Parse response
        return emptyList()
    }

    // MARK: - Helper Classes

    data class Session(
        val userId: String,
        val accessToken: String,
        val refreshToken: String
    ) {
        constructor(session: io.github.jan.supabase.gotrue.auth.Session) : this(
            userId = session.user.id,
            accessToken = session.accessToken,
            refreshToken = session.refreshToken ?: ""
        )
    }

    companion object {
        private const val TAG = "SupabaseService"
    }
}

